My topic 2

andrii-holovchenko opened the pull request.

andgol committed:
prog1

andgol committed:
prog2

yshestakov commented on prog1.py:
```diff
@@ -0,0 +1,28 @@
+# Detect local mounted disk
+local_disk_mounts = []
+partitions = psutil.disk_partitions(all=False)
+pattern = re.compile(".*/dev/[h,s]d[a-z].*")
```
Given regex doesn't handle local FSes created on LVM / disk-mapper devices like this:
```
/dev/mapper/rsws14_vg-lv_images on /images type ext4 (rw,relatime,data=ordered)
/dev/mapper/rsws14_vg-lv_docker_xfs on /var/lib/docker-bk type xfs (rw,relatime,attr2,inode64,noquota)
```

yshestakov commented on prog1.py:
```diff
@@ -0,0 +1,28 @@
+
+file_num = 3
+file_size = 4096
+#free_space = 
```
It's supported that parameters are specified by the user and passed to the program.
I suggest to use `argparse` Python module to handle parameters passed to the program:
https://docs.python.org/2/howto/argparse.html

yshestakov commented on prog1.py:
```diff
@@ -0,0 +1,28 @@
+
+
+start = time.time()
+run_dd(file_num, file_size, local_disk_mounts[0])
```
We have 2 parameters for the "populate disk" task:
- create Z files 
- ... of size Y,
with a special condition:
- run Z “dd” processes where each process will fill the selected file
I.e. Z "dd" processes need to be organized as process/thread pool and work in parallel, not sequentially

yshestakov commented on prog1.py:
```diff
@@ -0,0 +1,28 @@
+
+def run_dd(num, size, mount):
+    for n in range(num):
+        subprocess.check_output(['dd', 'if=/dev/zero', 'of=%sfile%s.txt' % (mount, n), 'count=1', 'bs=%s' % size ], stderr=subprocess.STDOUT)
```
1. No exit code check
2. Why do you redirect `stderr` to STDOUT?

yshestakov commented on prog1.py:
```diff
@@ -0,0 +1,28 @@
+partitions = psutil.disk_partitions(all=False)
+pattern = re.compile(".*/dev/[h,s]d[a-z].*")
+for p in partitions:
+   if pattern.match(str(p)) and psutil.disk_usage(p.mountpoint).free > 1024:
```
`1024` value there should be the `X` parameter: 
>  Detect local mounted disk (make sure it is local) with at least X MB free space

Next, what is the unit of space returned by `.free` properly: KB, MB, bytes?

yshestakov commented on prog2.py:
```diff
@@ -0,0 +1,38 @@
+        stdout=stdout.readlines()
+        for line in stdout:
+            print line
+        ssh.close()
```
1. I would suggest to prefix each line of output with hostname from which this output is received.
2. stderr is ignored. Why?
3. No exit code check.
4. Potential out-of-memory issue if executed command generates a lot of MB of data on stdout stream

yshestakov commented on prog2.py:
```diff
@@ -0,0 +1,38 @@
+import threading
+import time
+
+com = raw_input("Please specify the command to run: ")
```
I would suggest to used `argparse` module instead to parse command line parameters

yshestakov commented on prog2.py:
```diff
@@ -0,0 +1,38 @@
+    while count < len(hosts):
+            threading.Thread(target=run_ssh_command,args=(str(hosts[count]),com)).start()
+            time.sleep(0.5)
+            count+=1
```
What will happen if you have a large cluster of 1000 or 10000 nodes for example?
Would your system survive 1000 or 10000 threads/processed running in parallel?
I suggest to implement a thread/process pull:
   https://docs.python.org/2/library/multiprocessing.html

andgol committed:
Multiple improvements in prog1.py

andgol committed:
Multiple improvements in prog2.py

andgol committed:
Updated Readme

yshestakov commented on prog1.py:
```diff
@@ -1,28 +1,66 @@
-file_num = 3
-file_size = 4096
-#free_space = 
+args = parser.parse_args()
```
This call needs to be moved into `if __name__ == "__main__"` block, which is executed only when this Python module is running as a standalone program.
All statements in the "global" scope are executed immediately when the module is imported by `module_import`. It's not a critical issue, but a suggestion for a future.

yshestakov commented on prog1.py:
```diff
@@ -1,28 +1,66 @@
-start = time.time()
-run_dd(file_num, file_size, local_disk_mounts[0])
-end = time.time()
+    if args.number and args.size and args.mount_point:
```
1. The program needs to find a filesystem on the local disk. No need to specify mount point explicitly.
2. In case no free partitions (FS-es) are found -- you need to inform the user about this. Currently you simply ignores empty list returned by `list_disk_mounts()` function  as far as I See

yshestakov commented on prog1.py:
```diff
@@ -1,28 +1,66 @@
+                processes.append(p)
+                p.start()
+            for process in processes:
+                process.join()
```
Looks better. However, exit codes of background (dd) processes are not checked yet :)

andgol committed:
Aargs parser moved to main function;
Auto finding of available filesystems;
Check if no filesystems were found;
Exit code checking of background dd process

andgol committed:
Partitions detecting change
Logic restructured with adjustment of pyudev module

andrii-holovchenko committed:
Update README.md

Commit:
Reverted back to using regex for partitions finding

Commit:
Added \'os.path.join()'\ function for correctly consructing path names

Commit:
Fixed time tracking
